File: Recursion/Palindrome/Palindrome.cpp
#include<iostream>
using namespace std;

bool isPalindromeHelper(int l, int r, string &s) {

    if (l >= r) {
        return true;
    }
    if (s[l] != s[r]) {
        return false;
    }

    return isPalindromeHelper(l+1, r-1, s);
}

bool isPalindrome(string s) {

    string str = "";
    for (char c: s) {
        
        if((isalnum(c) == false) || (c == ' ')) {
            continue;
        }

        if (isupper(c)){
            c += 32;
        }

        str.push_back(c);
    }

    if(str.size() == 0)
        return true ;
    return isPalindromeHelper(0 , str.size() - 1 , str) ;
}

int main() {

    string s;
    cout<<"Enter the string: ";
    getline (cin, s);
    if (isPalindrome(s)) {
        cout<<"Its Palindrome";
    } else {
        cout<<"Not a palindrome";
    }
    return 0;
}

File: Recursion/Palindrome/Palindrome.py
class Palindrome:

    def isPalindrome(self, s: str) -> bool:

        def isPalindromeHelper(l, r, s):

            if l >= r:
                return True
            if s[l] != s[r]:
                return False
            return isPalindromeHelper(l+1, r-1, s)

        str = ""
        for char in s:
            if char.isalnum() == False or char == " ":
                continue
            if char.isupper():
                char = char.lower()
            str += char
        return isPalindromeHelper(0, len(str)-1, str)

str = input("Enter the string: ")
solution = Palindrome()
if (solution.isPalindrome(str)): 
    print("Its palindrome")
else:
    print("Its not palindrome")

    # time complexity: O(n)
    # space complexity: O(n)